### PR TITLE
gh-1285 Clean up deleted doc ids from inverted indices periodically

### PR DESCRIPTION
**Pull request by antas-marcin**

### CODE REVIEW

File: adapters/repos/db/shard.go [hunk 1]
```diff
@@ -191,3 +196,50 @@ func (s *Shard) addProperty(ctx context.Context, prop *models.Property) error {
...
+			return nil
+		}
+
+		err := docIDs.ForEach(func(documentID, v []byte) error {
+			lookup, err := docid.LookupFromBinary(v)
+			if err != nil {
+				return errors.Wrap(err, "lookup from binary")
+			}
+			if lookup.Deleted {
+				s.deletedDocIDs.Add(binary.LittleEndian.Uint32(documentID[4:]))
```
[the remark this answers is not in the document]
Comment (etiennedi) [reply adapters/repos/db/shard.go:213]: as discussed on slack, the root cause for this is outside of this issue, cleanup has been moved to https://github.com/semi-technologies/weaviate/issues/1282

File: adapters/repos/db/inverted/cleaner.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+func (c *Cleaner) getDocumentKey(documentID uint32) []byte {
+	keyBuf := bytes.NewBuffer(make([]byte, 4))
+	binary.Write(keyBuf, binary.LittleEndian, &documentID)
+	key := keyBuf.Bytes()
+	return key
+}
+
+func (c *Cleaner) propHasFrequency(p *models.Property) bool {
+	for i := range p.DataType {
+		if schema.DataType(p.DataType[i]) == schema.DataTypeString || schema.DataType(p.DataType[i]) == schema.DataTypeText {
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/inverted/cleaner.go:47]: Nice one! I really like this one :)

File: adapters/repos/db/shard.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/shard.go:213]: I have just marked this place with a comment so we will not forget about this place

File: adapters/repos/db/shard.go [hunk 3]
```diff
@@ -191,3 +196,50 @@ func (s *Shard) addProperty(ctx context.Context, prop *models.Property) error {
...
+		return errors.Wrap(err, "find deleted ids")
+	}
+
+	return nil
+}
+
+func (s *Shard) startPeriodicCleanup() error {
+	t := time.Tick(s.cleanupInterval)
+	batchCleanupInterval := 5 * time.Second
+	batchSize := 10
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/shard.go:233]: I have changed that default batch size from 10 to 100, will do further tests to define the best batchSize value

File: adapters/repos/db/inverted/cleaner.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	err := propsBucket.ForEach(func(item, data []byte) error {
+		return c.deleteFn(propsBucket, Countable{Data: item}, c.deletedDocIDs, hasFrequency)
+	})
+	if err != nil {
+		return errors.Wrapf(err, "cleanup property %s row", p.Name)
+	}
+	return nil
+}
+
+func (c *Cleaner) deleteDocument(tx *bolt.Tx, documentID uint32) bool {
```
[the remark this answers is not in the document]
Comment (antas-marcin) [reply adapters/repos/db/inverted/cleaner.go:70]: This actually deletes the document from the `DocIDBucket` so I think we should keep the name

**Review by etiennedi — APPROVED** [review]

**Pull request merged**